File: backend/app/config/features.py
"""
Feature flags for modular architecture.
Enables/disables features without code changes.
"""
from pydantic import BaseModel
from typing import Literal


class FeatureFlags(BaseModel):
    """Feature flags configuration"""

    # Map features
    enable_street_labels_toggle: bool = True
    default_map_mode: Literal["training", "reference"] = "training"
    enable_gis_layers: bool = True
    enable_fire_box_layer: bool = True
    default_show_fire_boxes: bool = False  # Start hidden for training mode

    # Quiz modes (can be disabled individually)
    quiz_beat_identification: bool = True
    quiz_facility_location: bool = True
    quiz_protocol_destination: bool = True
    quiz_turn_by_turn: bool = True

    # Data sources
    enable_moco_gis_sync: bool = True
    enable_osm_supplement: bool = True
    enable_hhs_nursing_homes: bool = True

    # Advanced features
    enable_spaced_repetition: bool = True
    enable_protocol_engine: bool = True
    enable_routing: bool = True

    # UI features
    enable_dark_mode: bool = True
    enable_dashboard: bool = True
    enable_progress_tracking: bool = True

    # Facility layers
    enable_hospitals_layer: bool = True
    enable_stations_layer: bool = True
    enable_nursing_homes_layer: bool = True

    class Config:
        """Pydantic configuration"""
        validate_assignment = True


# Global feature flags instance
features = FeatureFlags()


def get_features() -> dict:
    """
    Get feature flags as dictionary for API responses.

    Returns:
        dict: Feature flags configuration
    """
    return features.model_dump()

File: frontend/src/config/features.ts
/**
 * Feature flags interface - matches backend configuration
 */
export interface FeatureFlags {
  // Map features
  enableStreetLabelsToggle: boolean;
  defaultMapMode: 'training' | 'reference';
  enableGisLayers: boolean;
  enableFireBoxLayer: boolean;
  defaultShowFireBoxes: boolean;

  // Quiz modes
  quizBeatIdentification: boolean;
  quizFacilityLocation: boolean;
  quizProtocolDestination: boolean;
  quizTurnByTurn: boolean;

  // Data sources
  enableMocoGisSync: boolean;
  enableOsmSupplement: boolean;
  enableHhsNursingHomes: boolean;

  // Advanced features
  enableSpacedRepetition: boolean;
  enableProtocolEngine: boolean;
  enableRouting: boolean;

  // UI features
  enableDarkMode: boolean;
  enableDashboard: boolean;
  enableProgressTracking: boolean;
}

/**
 * Default feature flags (fallback if API fails)
 */
export const defaultFeatures: FeatureFlags = {
  enableStreetLabelsToggle: true,
  defaultMapMode: 'training',
  enableGisLayers: true,
  enableFireBoxLayer: true,
  defaultShowFireBoxes: false,
  quizBeatIdentification: true,
  quizFacilityLocation: true,
  quizProtocolDestination: true,
  quizTurnByTurn: true,
  enableMocoGisSync: true,
  enableOsmSupplement: true,
  enableHhsNursingHomes: true,
  enableSpacedRepetition: true,
  enableProtocolEngine: true,
  enableRouting: true,
  enableDarkMode: false,
  enableDashboard: true,
  enableProgressTracking: true,
};

/**
 * Convert snake_case backend keys to camelCase frontend keys
 */
function convertToCamelCase(obj: any): FeatureFlags {
  return {
    enableStreetLabelsToggle: obj.enable_street_labels_toggle,
    defaultMapMode: obj.default_map_mode,
    enableGisLayers: obj.enable_gis_layers,
    enableFireBoxLayer: obj.enable_fire_box_layer,
    defaultShowFireBoxes: obj.default_show_fire_boxes,
    quizBeatIdentification: obj.quiz_beat_identification,
    quizFacilityLocation: obj.quiz_facility_location,
    quizProtocolDestination: obj.quiz_protocol_destination,
    quizTurnByTurn: obj.quiz_turn_by_turn,
    enableMocoGisSync: obj.enable_moco_gis_sync,
    enableOsmSupplement: obj.enable_osm_supplement,
    enableHhsNursingHomes: obj.enable_hhs_nursing_homes,
    enableSpacedRepetition: obj.enable_spaced_repetition,
    enableProtocolEngine: obj.enable_protocol_engine,
    enableRouting: obj.enable_routing,
    enableDarkMode: obj.enable_dark_mode,
    enableDashboard: obj.enable_dashboard,
    enableProgressTracking: obj.enable_progress_tracking,
  };
}

/**
 * Fetch feature flags from backend API
 */
export async function fetchFeatures(apiBaseUrl: string): Promise<FeatureFlags> {
  try {
    const response = await fetch(`${apiBaseUrl}/api/features`);
    if (!response.ok) {
      console.warn('Failed to fetch features, using defaults');
      return defaultFeatures;
    }
    const data = await response.json();
    return convertToCamelCase(data);
  } catch (error) {
    console.error('Error fetching features:', error);
    return defaultFeatures;
  }
}
